src/webui/gitutils.jl

```diff
@@ -0,0 +1,161 @@
+    else
+        # First check for organization membership, and fall back to collaborator status.
+        ismember = @gf is_member(u.forge, repo.organization.login, u.user.login)
+        something(ismember, false) ||
+            @gf is_collaborator(u.forge, repo.organization.login, repo.name, u.user.login)
+    end
+    return something(hasauth, false)
+end
+
+function isauthorized(u::User{GitLab.User}, repo::GitLab.Project)
+    repo.visibility == "private" && return false
+    forge = PROVIDERS["gitlab"].client
+    hasauth = if repo.namespace.kind == "user"
+        @gf is_collaborator(forge, repo.owner.username, repo.name, u.user.id)
+    else
+        # Same as above: group membership then collaborator check.
+        ismember = @gf is_member(forge, repo.namespace.full_path, u.user.id)
+        something(ismember, false) ||
+            @gf is_collaborator(u.forge, repo.organization.login, repo.name, u.user.id)
+    end
+    return something(hasauth, false)
+end
+
+# Get the raw Project.toml text from a repository.
+function gettoml(f::GitHubAPI, repo::GitHub.Repo, ref::AbstractString)
+    fc = @gf get_file_contents(f, repo.owner.login, repo.name, "Project.toml"; ref=ref)
+    return fc === nothing ? nothing : String(base64decode(strip(fc.content)))
+end
+
+function gettoml(::GitLabAPI, repo::GitLab.Project, ref::AbstractString)
+    forge = PROVIDERS["gitlab"].client
+    fc = @gf get_file_contents(forge, repo.id, "Project.toml"; ref=ref)
+    return fc === nothing ? nothing : String(base64decode(fc.content))
+end
+
+function getcommithash(f::GitHubAPI, repo::GitHub.Repo, ref::AbstractString)
+    commit = @gf get_commit(f, repo.owner.login, repo.name, ref)
+    return commit === nothing ? nothing : commit.sha
+end
+
+function getcommithash(f::GitLabAPI, repo::GitLab.Project, ref::AbstractString)
+    forge = PROVIDERS["gitlab"].client
+    commit = @gf get_commit(forge, repo.id, ref)
+    return commit === nothing ? nothing : commit.id
+end
+
+# Get a repo's clone URL.
+cloneurl(r::GitHub.Repo) = r.clone_url
+cloneurl(r::GitLab.Project) = r.http_url_to_repo
+
+# Get a repo's tree hash.
+function gettreesha(f::GitHubAPI, r::GitHub.Repo, ref::AbstractString)
+    branch = @gf get_branch(f, r.owner.login, r.name, ref)
+    return branch === nothing ? nothing : branch.commit.commit.tree.sha
+end
+
+function gettreesha(::GitLabAPI, r::GitLab.Project, ref::AbstractString)
+    url = cloneurl(r)
+
+    if REGISTRY[] isa Registry{GitLabAPI}
+        # For private repositories, we need to insert the token into the URL.
+        host = HTTP.URI(url).host
+        token = CONFIG["gitlab"]["token"]
+        url = replace(url, host => "oauth2:$token@$host")
+    end
+
+    return try
+        mktempdir() do dir
+            dest = joinpath(dir, r.name)
+            run(`git clone $url $dest`)
+            match(r"tree (.*)", readchomp(`git -C $dest show $ref --format=raw`))[1]
+        end
+    catch ex
+        println(get_backtrace(ex))
```

Comment:
Pretty sure this doesn't actually display the exception, only the backtrace.
Why not do the conventional `@error "This thing messed up" exception=(ex, catch_backtrace())` ? 
This is the same question that I had previously.

Reply:
And I have already answered earlier that `Logging` does not print the stacktrace in a readable format. It prints it without the newlines. If you are using a better logger please let me know or add it to the Project file.

Reply:
You can just change `SimpleLogger(io)` for `ConsoleLogger(io)` (from the Logging stdlib)
and you get Logging that looks just like the repl.

If color ends up being a problem (I can't recall if it naturally strips color if write to a stream,
then pass the kwarg:
```
meta_formatter = (args...) -> :normal
```
to the constructor